### ISSUE TITLE
Support onDidExecuteCommand

### ISSUE THREAD

**Issue by waderyan** [microsoft/vscode#1431]
I am building an extension for VS Code. My extension will show the key bindings associated to a recently action (I am a fan of this for helping me learn key bindings).

Looking through the documentation there seems to be a way to access the commands in the editor, but no way that I can find to listen for actions / commands as they are invoked.

I would like the ability to listen for all user actions and then react to their actions.

**Comment by egamma** [comment]
This is currently supported by calling this method:
vscode.commands.getCommands(filterInternal?: boolean): Thenable<string[]>;

Does this unblock you?

**Comment by wadeanderson7** [comment]
Not quite. From what I can tell in the docs, this function will return a list of commands. What I want is to listen to commands as they happen. 

My specific use case is:
- Listen for commands
- When command occurs, determine if their is a key binding for that command
- Display in the status bar (or some UI element) the key binding for the command the user just executed.

**Comment by egamma** [comment]
Got it, this is currently not supported. 
@jrieken fyi.

**Comment by siegebell** [comment]
I would also like to see this feature.

I am writing an extension that needs to monitor cursor movements and make slight cursor position adjustments on occasion. The only way I've found to do this is to override the keybinding for e.g. 'left', call 'cursorLeft' from my command, and then validate the cursor's position.

Even without validation, intercepting 'left' and then immediately calling 'cursorLeft' from an extension causes the UI to be *much less responsive* than binding directly 'left' to 'cursorLeft'. (Because extensions are given low priority.)

**Comment by jrieken** [comment]
@johnfn for the VIM extension do you need `onDidExecuteCommand`, or `onWillExecuteCommand`, or both? The first (`onDidExecuteCommand`) is fairly easy to implement but the latter raises many questions...

**Comment by johnfn** [comment]
@jrieken I assume this is for tab complete? If so, post-trigger would be the most useful.

**Comment by rebornix** [comment]
@jrieken we will benefit a lot with `onDidExecuteCommand`, for example, code snippet then we know how to handle those selections.

Like @johnfn mentioned, `onWillExecuteCommand` might be useful for Tab complete, etc. I'll put more if I come up with more usage.

**Comment by jrieken** [comment]
@johnfn @rebornix  not quite sure what you guys mean by 'Tab complete'?

**Comment by rebornix** [comment]
@jrieken A common user case is like:

1. User types `ab` in a TypeScript file
2. Code will suggest `abstract`
3. User press `<Tab>`, Code will do autocomplete and we will get `abstract`.
4. User press `.` to repeat previous action.

Since the only user does is pressing  `a`, `b`, `Tab` but we have no idea about the final result of these three keystrokes. so when we repeat previous action, the result will be just another `ab`, which is not what we want. The reason that we don't get `<Tab>` is we are listening to `type` command and `<Tab>` (Up, Down, Left, Right, etc) will not be passed to us.

So if we can bind handlers to `onDidExecuteCommand` and we know it's an autocomplete, maybe we can put it in our cached history and repeat correctly?

**Comment by alexandrudima** [comment]
@rebornix I am not convinced that "spying" on commands will help you in any meaningful way. For example, in the example you make, where a user types `ab` followed by `<Tab>`, you will find out the command `acceptSelectedSuggestion` has executed. But IMHO this does not help you to "repeat" the action, as the command `acceptSelectedSuggestion` is for example completely contextual. Perhaps the suggest widget has even closed in the meantime, so executing it programmatically will result in a no-op.

Perhaps `.` can be supported only for vim commands rather than for any command plugged into vscode? Would that be a vim adoption blocker?

**Comment by rebornix** [comment]
@alexandrudima Thanks for your input. I agree that `acceptSelectedSuggestion` execution doesn't solve this problem. But since we are only monitoring `type` command then we have no idea what's the exact content being inserted into the editor, we can't implement `.` perfectly.

`.` is a Vim adoption blocker as well as `u` (undo). Since you are enriching Code's `undo`/`redo` commands, will there be any additional info to help implement `repeat`?

**Comment by alexandrudima** [comment]
@rebornix My question stands: "Perhaps `.` can be supported only for vim commands rather than any command plugged into vscode?".

There are a lot of commands in VSCode out of the box and even more from extensions. Types of commands that I can think of:
* commands that edit the document: e.g. toggle line comment (`ctrl+/` or `cmd+/`)
* commands that change the editor "view" state: e.g. actions.find (`ctrl+f` or `cmd+f`), find all references, show hover, toggle word wrap
* commands that change the workbench state: e.g. open markdown preview to the side, show references, go to definition, hide sidebar

There are also edits to the document that don't stem from a command (e.g. format on type)

The commands are not categorized statically in any meaningful way. They execute and they do something. Sometimes, that results in a document edit (that can be observed via `onDidChangeTextDocument`). 

Also note the `textEditor.edit` API that is currently in place. The only information I get from that are the edits to be aplied to a text document, which are immediately reflected in `onDidChangeTextDocument`.

Is it possible to point me to an example/explanation of how `.` should work in Vim?

**Comment by johnfn** [comment]
The big thing we need is the ability to know that an action such as tab complete or snippet insert or any other sort of text change has taken place. If you guys can't figure out exactly how to let us repeat it, that's okay; we can just manually diff the document and figure out what happened (most of the time). 

Quick explanation for .: The idea is that it allows you to repeat an action somewhere else in the document. So if you inserted some text on the first line, then navigated to the last line, . would add the text down there. If you tab-completed a word on the first line and hit . on the last line, it'd insert the full word. etc, etc. `.` only concerns itself with text-related actions.

**Comment by jrieken** [comment]
> only concerns itself with text-related actions

Isn't that an argument for the [`TextDocumentChangeEvent`](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L3455)? We could maybe enrich it with a `source`-thing - similar to [`TextEditorSelectionChangeEvent#kind`](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L538). That would indicate (somehow?) that this textual change was done because of a command and the `.`-feature would then remember it for repeated execution

**Comment by alexandrudima** [comment]
@jrieken :+1: However, the source is currently a string (not an enum) and at least for edits coming from the extension host the source is set to `MainThreadTextEditor` -- see https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/mainThreadEditorsTracker.ts#L342

**Comment by jrieken** [comment]
As long as the strings being sort well defined we can do something like [this](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostTypes.ts#L839)

**Comment by jrieken** [comment]
@johnfn @rebornix I have extracted https://github.com/Microsoft/vscode/issues/10801 for the last bit of this discussion since this issue is still about events for command execution.

**Comment by mimse** [comment]
I would like to see this feature implemented as well

**Comment by jrieken** [comment]
> This feature request will not be considered in the next [6-12 months roadmap](https://github.com/Microsoft/vscode/wiki/Roadmap) and as such will be closed to keep the number of issues we have to maintain actionable. Thanks for understanding and happy coding!

**Comment by hedgerh** [comment]
@jrieken would PRs be accepted on this, by any chance?

**Comment by jrieken** [comment]
Looks like you are planning to use this for marcos... Assuming you have this API can you sketch out how you would use it? Did you consider that there are commands in the UX side and extension host. Also, commands sometimes carry arguments along and often virtual commands are created when invoking an extension-host-side command from the UX-side (to prevent arguments from travelling between the processes)...

**Comment by hedgerh** [comment]
My original thought was to be able to listen using `vscode.commands.onDidExecuteCommand(cb)` in an extension and be able to catch UX events, but I've been playing around in the codebase to see how it could be implemented, and realized it wasn't as simple as I hoped.  I couldn't just put an `Emitter` on `CommandService`, pass a cb from the extension host, and have the cb invoked when commands fired on the editor side.

So I guess it's not feasible/possible to listen for commands from the UX side because it's on a separate process?  Learning a bit about the codebase was interesting, at least.

**Comment by hedgerh** [comment]
Suppose I did have the working API with UX commands, it may look something like this:

- User starts recording with a "Start Recording" command from the command palette.
- Extension listens for commands.  On execute, the command ids are pushed to an array.
- User stops recording.  The macro is written to `macros` in their global user settings.
- The `macros` extension would handle the actual execution of macros.

**Comment by hedgerh** [comment]
I managed to get it working by having an Emitter on `extHostCommands` that holds the actual subscribe callbacks, then another Emitter on `commandService` that `mainThreadCommands` subscribes to.  The `mainThreadCommands` sends the command payload over to the extension host, instead of trying to fire a callback.

**Comment by jrieken** [comment]
Good progress, remains how to handle command arguments (that sometimes cannot be send across the wire) and how to handle internal commands. The event could ignore them, we usually prefix them with `_` or you planned extension ignores them. However, how to handle arguments isn't really clear to me yet...

**Comment by hedgerh** [comment]
Can you tell me a little more about these command arguments? Why can't they be sent across the wire? Do you think any of them would be something an extension would be interested in?  Any possible workarounds?  If they can't be transmitted, it may just be noted that extensions can see the commands firing, but won't be able to access those arguments.

As for internal commands, we could filter them by default if the extension most likely won't be interested in them, maybe provide a flag to enable them (or not.)

Does my approach so far seem reasonable so far?

**Comment by jrieken** [comment]
A simple sample is the `type` command - as it name suggests it happens when you type in the editor and the argument is the actual character/characters that are being typed. More complex examples are things like 'git stage'. It gets called with the resource that's being staged. Because that cannot be send we create a fake (indirection) command that being invoked just to invoke another command. The challenge will be to filter those indirection commands.

**Comment by hedgerh** [comment]
Not sure if I'm missing something, but I'm able to receive the `type` command with an arg that's a 1 element array containing the key pressed.  Don't see any weird commands coming through.  I'll try to do `git stage` and see what comes out.  Anything I should be looking for in the code to identify these fake commands?  Thanks for your help, by the way.

**Comment by jrieken** [comment]
Most (not sure if all) should be named [`_internal_command_delegation_TIMESTAMP`](https://github.com/Microsoft/vscode/blob/7ffcfd503d6c86b085005abb08fbba9a183c2696/src/vs/workbench/api/common/extHostCommands.ts#L208), so in theory ignore `_`'ed command should work

**Comment by hedgerh** [comment]
Pushed up my work and opened up a RFC/WIP PR of what I came up with.  Interesting enough, none of the internal commands actually go through the event emitter in the command service, so they don't surface on the extension side.  I verified that internal commands were firing, just not through the emitters.

Interested if there are any performance implications with talking between the two processes like this.  Possible enhancement would be to allow a glob/regex/array of strings/wildcard as the first param that would act as a filter for what events surface to the extension.

**Comment by jrieken** [comment]
This PR is merged but there is some little (design) work left

- [ ] decide if we should hide private/underscore commands
- [ ] deal with the internal command delegation commands
- [ ] add some more test

**Comment by jrieken** [comment]
Closing this for July testing

**Comment by alexr00** [comment]
This seems to work well, but I'm seeing the event get triggered twice when running this hello world command:
```typescript
export function activate(context: vscode.ExtensionContext) {
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "helloworld-sample" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	const disposables: vscode.Disposable[] = [];
	disposables.push(vscode.commands.registerCommand('extension.helloWorld', () => {
		// The code you place here will be executed every time your command is executed

		// Display a message box to the user
		vscode.window.showInformationMessage('Hello World!');
	}));

	disposables.push(vscode.commands.onDidExecuteCommand((e: vscode.CommandExecutionEvent) => {
		console.log(e.command);
	}));

	context.subscriptions.push(...disposables);
}
```
The command isn't actually running twice, but the listener for `onDidExecuteCommand` does get called twice.

**Comment by jrieken** [comment]
We have decided against this API: https://github.com/microsoft/vscode/issues/78091#issuecomment-518545491
